docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+
+```purs
+-- Before
+foo = 1 + 2
+  --  ^^^^^ Rewrite rule: "I can optimize that! `1 + 2` is `3`"
+
+-- After
+foo = 3
+```
+
+However, some expressions will prevent such rewrite rules from triggering. For example:
+
+```purs
+foo =
+  let
+    a = { bar: 42 }
+  in
+    1 + a.bar
+--  ^^^^^^^^^ Rewrite rule: "I can't optimize that! `a.bar` is not a literal `Int` value."
+```
+
+Because the rewrite rule cannot see the literal `Int` value represented by `a.bar`, its optimization does not trigger. To solve this problem, we need to make it visible. The solution is to duplicate code via inlining.
+
+**To summarize, inlining duplicates code so that rewrite rules can trigger optimizations that were otherwise hidden from its eyes.**
+
+#### Code Optimzation Example via Rewrite Rules and Inlining
+
+Let's see how this works in practice using this example:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = ignoreArgs 8 9
+```
+
+Let's assume `ignoreArgs` has an inline directive of `arity=1`. Here's what happens when we optimize this code piece.
+
+First, we see that `ignoreArgs` has 2 arguments applied. Since `2` >= `1`, the arity of the directive we specified, we inline `ignoreArgs`. Since `8` corresponds to `arg1`, we replace `arg1` with `8` when inlining `ignoreArgs`. We don't remove the original `ignoreArgs` declaration.
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = (\arg2 -> do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 8 + arg2
+  ) 9
+```
+
+At this point, `purs-backend-es`' default inliners will see a lambda (i.e. a "producer") being immediately applied to an argument (i.e. a "consumer"). Thus, it will inline that argument into the lambda's body. This gets us:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 8 + 9
+```
+
+A rewrite rule will detect the `8 + 9` expression, and replace it with its sum: `17`. However, `1 + a.bar` and `a.bar + 17` don't trigger the rewrite rule because one of the values is not a literal `Int` value.
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + 17
+```
+
+Fortunately, `purs-backend-es`' default inliners will see a record binding (i.e. a "producer") that is immediately accessed under the label `bar`. Thus, the corresponding value will be inlined:
+
+```purs
+ignoreArgs arg1 arg2 = do
+  let
+    a = { bar: 41 }
+  1 + a.bar + arg1 + arg2
+
+foo = do
+  let
+    a = { bar: 41 }
+  1 + 41 + 17
+```
```

Comment:
This is inaccurate. The optimizer uses analysis to determine whether or not a binding should be inlined, and it will do that in lieu of creating a binding. So in this case, for `ignoreArgs`, what it actually does is: 
```
ignoreArgs arg1 arg2 = do
  let
    a = { bar: 41 }
  1 + a.bar + arg1 + arg2
```
`a` is a record constructor and is eliminated in all references, so it can be inlined.
```
ignoreArgs arg1 arg2 = do
  1 + { bar: 42 }.bar + arg1 + arg2
```
The record constructor is immediately eliminated, so it can be reduced:
```
ignoreArgs arg1 arg2 = do
  1 + 42 + arg1 + arg2
```
The operands of + are both known literals (up to some fixed depth, associatively), so it can be reduced.
```
ignoreArgs arg1 arg2 = do
  43 + arg1 + arg2
```